homeassistant/components/evohome/__init__.py

```diff
@@ -221,9 +251,116 @@ def _handle_exception(err) -> bool:
         broker.update, config[DOMAIN][CONF_SCAN_INTERVAL]
     )
 
+    setup_service_functions(hass, broker)
+
     return True
 
 
+def setup_service_functions(hass: HomeAssistantType, broker):
+    """Set up the service functions.
+
+    Not all evohome-compatible systems support all system modes.
+    """
+
+    async def force_refresh(call):
+        """Obtain the latest state data via the vendor's RESTful API."""
+        await broker.update()
+
+    async def set_system_mode(call):
+        """Set the system mode."""
+        payload = {
+            "unique_id": broker.tcs.systemId,
+            "service": call.service,
+            "data": call.data,
+        }
+        hass.helpers.dispatcher.async_dispatcher_send(DOMAIN, payload)
+
+    async def set_zone_override(call):
+        """Set the zone override (setpoint)."""
+        entity_id = call.data[ATTR_ENTITY_ID]
+
+        registry = await async_get_registry(hass)
+        registry_entry = registry.async_get(entity_id)
+
+        if registry_entry is None or registry_entry.platform != DOMAIN:
+            raise KeyError(f"'{entity_id}' is not a known evohome entity")
+
+        if registry_entry.domain != "climate":
+            raise KeyError(f"'{entity_id}' is not an evohome zone")
+
+        payload = {
+            "unique_id": registry_entry.unique_id,
+            "service": call.service,
+            "data": call.data,
+        }
+
+        hass.helpers.dispatcher.async_dispatcher_send(DOMAIN, payload)
+
+    hass.services.async_register(
+        DOMAIN,
+        SVC_SET_ZONE_OVERRIDE,
+        set_zone_override,
+        schema=SET_ZONE_OVERRIDE_SCHEMA,
+    )
+    hass.services.async_register(
+        DOMAIN,
+        SVC_RESET_ZONE_OVERRIDE,
+        set_zone_override,
+        schema=RESET_ZONE_OVERRIDE_SCHEMA,
+    )
+
+    modes = broker.config["allowedSystemModes"]
+
+    if [m["systemMode"] for m in modes if m["systemMode"] == "AutoWithReset"]:
+        hass.services.async_register(
+            DOMAIN, SVC_RESET_SYSTEM, set_system_mode, schema=vol.Schema({})
+        )
+
+    system_mode_schemas = []
+    modes = [m for m in modes if m["systemMode"] != "AutoWithReset"]
+
+    perm_modes = [m["systemMode"] for m in modes if not m["canBeTemporary"]]
```

Comment:
Please add some more documentation what is going on. I was able to follow it but I had to study the code.

Reply:
Yes, I agree this was a pain: added, hope I haven't gone overboard.